pandas/indexes/base.py

```diff
@@ -1923,14 +1924,38 @@ def difference(self, other):
 
         other, result_name = self._convert_can_do_setop(other)
 
-        theDiff = sorted(set(self) - set(other))
-        return Index(theDiff, name=result_name)
+        this = self._get_unique_index()
+
+        # In what follows, get_indexer doesn't treat NaN's specially,
+        # so it would break the existing behavior, e.g.:
+        # `Index([1, nan]).difference(Index[nan]) == Index([nan, 1])`.
+        # Thus, dropping NaN's from `other` is a hack for backward compat.
+        # We do it here instead of manipulating `the_diff` later.
+        # (Check for MultiIndex is to get around a .hasnans exception)
+        dropna = not isinstance(self, ABCMultiIndex) and \
+            not isinstance(other, ABCMultiIndex) and \
+            self.hasnans and other.hasnans
+        other = other._get_unique_index(dropna=dropna)
```

Comment:
Removing `dropna` would make it treat NaN'a as any other element, so the following would be true
```python
pd.Index([1, np.nan]).difference(pd.Index([2, np.nan])) == pd.Index([1])
```
Similarly for `symmetric_difference`. It would simplify the code and be consistent with `intersection`. So, I'm for it.

It would break one test that checks for nan's in `symmetric_difference` https://github.com/pydata/pandas/blob/master/pandas/tests/indexes/test_base.py#L771
(related to #6444 - sorting of nan's).